### PR TITLE
CB-251: Reworked WS.java

### PR DESCRIPTION
**Pull request by jmittert**
* Added support for sending JSON requests
* Updated to use the latest Apache httpclient functions
* The send body is no longer always null
* Now handles headers properly

Test run against grimoire with 14 failures, production has 24.
http://ci/view/Voodoo/job/Voodoo_CB_Custom_Tests/88/
http://ci/view/Voodoo/job/Voodoo_CB_Custom_Tests/89/

http://ci/job/7700beta6_Voodoo_Self_Tests/28/

One failure was due to a chromedriver failure, the other due to an unrelated issue.

### CODE REVIEW

File: src/main/java/com/sugarcrm/candybean/webservices/WS.java [hunk 1]
```diff
@@ -21,216 +21,239 @@
...
+		HashMap<String, String> headers = new HashMap<>();
+		for (HashMap<String, String> map : postHeaders) {
+			for (Map.Entry<String, String> entry : map.entrySet()) {
+				headers.put(entry.getKey(), entry.getValue());
+			}
+		}
+		if (body == null) {
+			return request(op, uri, headers, payload);
+		} else {
+			return request(op,uri, headers, body, ContentType.DEFAULT_TEXT);
```
Comment (Eriten) [inline src/main/java/com/sugarcrm/candybean/webservices/WS.java]: Inconsistent format! space after comma

File: src/main/java/com/sugarcrm/candybean/webservices/WS.java [hunk 2]
```diff
@@ -21,216 +21,239 @@
...
+	 * Send an DELETE, GET, POST, or PUT http request using a JSON body
+	 * @param op The type of http request
+	 * @param uri The http endpoint
+	 * @param headers Map of header key value pairs
+	 * @param body Map of Key Value pairs
+	 * @return Key Value pairs of the response
+	 * @throws Exception If HTTP request failed
+	 */
+	public static Map<String, Object> request(OP op, String uri, Map<String, String> headers, Map<String,String> body) throws Exception {
+		return request(op,uri,headers,body,ContentType.APPLICATION_JSON);
```
Comment (Eriten) [inline src/main/java/com/sugarcrm/candybean/webservices/WS.java]: Inconsistent format! space after comma

**Comment by jmittert** [comment]
Cleaned up several whitespace issues.

File: src/main/java/com/sugarcrm/candybean/webservices/WS.java [hunk 3]
```diff
@@ -5,232 +5,251 @@
...
+					} else {
+						JSONObject jsonBody = new JSONObject(body);
+						StringEntity strBody = new StringEntity(jsonBody.toJSONString(), ContentType.APPLICATION_JSON);
+						put.setEntity(strBody);
+					}
 				}
-				request = put;
-	            mapParse = requestIt(request, headers);
-				break;
+				return handleRequest(put, headers);
```
Comment (Eriten) [inline src/main/java/com/sugarcrm/candybean/webservices/WS.java]: The PUT and POST pretty much have the same code inside the case. We should refactor this. Put the code out and put it in a private method and you can represent `HttpPut` and `HttpPost` as `HttpEntityEnclosingRequestBase`.

File: src/main/java/com/sugarcrm/candybean/webservices/WS.java [hunk 4]
```diff
@@ -5,232 +5,251 @@
...
+	 * Send a DELETE, GET, POST, or PUT http request, intended for multipart data and JSON requests
+	 * @param op The type of http request
+	 * @param uri The http endpoint
+	 * @param headers Map of header key value pairs
+	 * @param body Map of Key Value pairs
+	 * @param contentType The intended content type of the body
+	 * @return Key Value pairs of the response
+	 * @throws Exception If HTTP request failed
+	 */
+	public static Map<String, Object> request(OP op, String uri, Map<String, String> headers, Map<String, Object> body, ContentType contentType) throws Exception {
```
Comment (Eriten) [inline src/main/java/com/sugarcrm/candybean/webservices/WS.java]: Do a type check for Object in the body

Comment (Eriten) [inline src/main/java/com/sugarcrm/candybean/webservices/WS.java]: Let's write some documentation about what are the types of Map is accepted for body.

**Comment by jmittert** [comment]
- Added a method to handle both POST and PUT requests
- Added documenation about the Map<String,Object> type
- Verified the JSON Map before using it
- Added a unit test for the verification function
- Whitespace fixes

File: src/test/java/com/sugarcrm/candybean/webservices/WSSystemTest.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		nestedJSONData2.put("key6", "value6");
+		jsonData.put("key2", nestedJSONData1);
+		jsonData.put("key3", nestedJSONData2);
+		try {
+			headers.put("Content-Type", "application/json");
+			response = WS.request(WS.OP.POST, uri + "/post", headers, jsonData, ContentType.APPLICATION_JSON);
+		} catch (Exception e) {
+			Assert.fail(e.toString());
+		}
+		System.out.println(response.get("json"));
```
Comment (Eriten) [inline src/test/java/com/sugarcrm/candybean/webservices/WSSystemTest.java]: I think that this line is no longer needed.

**Comment by Eriten** [comment]
If you have time, just remove the `System.out.println`. Other than that, I think the PR looks good to me.

**Comment by jmittert** [comment]
Removed.
Running tests here: http://ci/job/Voodoo_CB_Custom_Tests/97/ and here http://ci/job/Voodoo_CB_Custom_Tests/98/

**Pull request merged**